pkg/runtime/field_export_reconciler.go

```diff
@@ -0,0 +1,674 @@
+type fieldExportReconciler struct {
+	reconciler
+	// rd is only used if binding to an ACK resource (not `FieldExport`)
+	rd *acktypes.AWSResourceDescriptor
+}
+
+// BindControllerManager sets up the AWSResourceReconciler with an instance
+// of an upstream controller-runtime.Manager
+func (r *fieldExportReconciler) BindControllerManager(mgr ctrlrt.Manager) error {
+	r.kc = mgr.GetClient()
+	r.apiReader = mgr.GetAPIReader()
+
+	if r.rd != nil {
+		return errors.New("cannot bind field export reconciler with non-nil resource descriptor")
+	}
+
+	return ctrlrt.NewControllerManagedBy(
+		mgr,
+	).For(
+		// Read only field export objects
+		&ackv1alpha1.FieldExport{},
+	).WithEventFilter(
+		predicate.GenerationChangedPredicate{},
+	).Complete(r)
+}
+
+// BindServiceResourceToManager binds a given AWS resource descriptor to the
+// controller manager
+func (r *fieldExportReconciler) BindServiceResourceManager(mgr ctrlrt.Manager) error {
+	r.kc = mgr.GetClient()
+	r.apiReader = mgr.GetAPIReader()
+
+	if r.rd == nil {
+		return errors.New("cannot bind resource field export reconciler with nil resource descriptor")
+	}
+
+	return ctrlrt.NewControllerManagedBy(
+		mgr,
+	).For(
+		(*r.rd).EmptyRuntimeObject(),
+	).WithEventFilter(
+		// Update on both status and spec changes
+		predicate.ResourceVersionChangedPredicate{},
+	).Complete(r)
+}
+
+// Reconcile implements `controller-runtime.Reconciler` and handles reconciling
+// a CR CRUD request
+func (r *fieldExportReconciler) Reconcile(ctx context.Context, req ctrlrt.Request) (ctrlrt.Result, error) {
+	return r.handleReconcileError(r.reconcile(ctx, req))
+}
+
+func (r *fieldExportReconciler) reconcile(ctx context.Context, req ctrlrt.Request) error {
+	// Determine if we are reconciling an ACK resource
+	if r.rd != nil {
+		return r.reconcileSourceResource(ctx, req)
+	}
+
+	// We are reconciling a field export CR
+	return r.reconcileFieldExport(ctx, req)
+}
+
+// reconcileFieldExport handles updates to `FieldExport` resources
+func (r *fieldExportReconciler) reconcileFieldExport(ctx context.Context, req ctrlrt.Request) error {
+	feObject, err := r.getFieldExport(ctx, req)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// resource wasn't found. just ignore these.
+			return nil
+		}
+		return err
+	}
+
+	sourceGK := feObject.Spec.From.Resource.GroupKind
+	sourceName := types.NamespacedName{
+		Name: *feObject.Spec.From.Resource.Name,
+		// We only support pulling from resources in
+		// the same namespace
+		Namespace: req.Namespace,
+	}
+
+	// Check if the target API group matches with the controller
+	var controllerRMF acktypes.AWSResourceManagerFactory
+	for _, v := range r.sc.GetResourceManagerFactories() {
+		controllerRMF = v
+		break
+	}
+	if sourceGK.Group != controllerRMF.ResourceDescriptor().GroupKind().Group {
+		ackrtlog.DebugFieldExport(r.log, feObject, "target resource API group is not of this service. no-op")
+		return nil
+	}
+
+	if feObject.DeletionTimestamp != nil {
```

Comment:
I specifically put it after the group check because I only want the controller that manages FieldExports for this type to be able to clean it up. I realise now that someone could change the spec and the service identifier in the finalizer would not match the service in the spec. Probably need to rethink this logic a little bit.

Reply:
Okay so I've just added a check inside `markUnmanaged` that only attempts to remove the finalizer if it's detected (that is, the field export has previously been managed by the current controller). Then I moved this line to above the sourceGK check, so every reconciler will do this check. Essentially all controllers will see the deletion timestamp and attempt to remove their respective finalizers if they see it.